### PR TITLE
dFEM Boundary Integrator

### PR DESCRIPTION
**Pull request by jandrej**
Add implementation for `DifferentiableOperator::AddBoundaryIntegrator`

### CODE REVIEW

**Comment by tzanio** [comment]
This PR is now [under review](https://github.com/mfem/mfem/blob/master/CONTRIBUTING.md#pull-requests) (see the table in the PR description). To help with the review process, please **do not force push** to the branch.

File: fem/dfem/doperator.hpp [hunk 1]
```diff
@@ -477,32 +558,52 @@ void DifferentiableOperator::AddDomainIntegrator(
...
+      {
+         elem_attributes[i] = mesh.GetAttribute(i);
+      }
+   }
+   else if constexpr (std::is_same_v<entity_t, Entity::BoundaryElement>)
+   {
+      for (int i = 0; i < mesh.GetNBE(); ++i)
+      {
+         elem_attributes[i] = mesh.GetBdrAttribute(i);
+      }
```
Comment (helloworld922) [inline fem/dfem/doperator.hpp]: Perhaps something like:
```c++
   const Array<int> *elem_attributes = nullptr;
   if constexpr (std::is_same_v<entity_t, Entity::Element>)
   {
      elem_attributes = &mesh.GetElementAttributes();
   }
   else if constexpr (std::is_same_v<entity_t, Entity::BoundaryElement>)
   {
      // would have to add GetBdrElementAttributes to Mesh since this has different logic from Mesh::GetBdrFaceAttributes
      elem_attributes = &mesh.GetBdrElementAttributes();
   }
```

File: tests/unit/dfem/test_mass.cpp [hunk 2]
```diff
@@ -51,37 +44,100 @@ void DFemMass(const char *filename, int p, const int r)
...
+      fes.GetRestrictionMatrix()->Mult(x, X);
+      dop.Mult(X, Z);
+
+      Y -= Z;
+
+      real_t norm_g, norm_l = Y.Normlinf();
+      MPI_Allreduce(&norm_l, &norm_g, 1, MPI_DOUBLE, MPI_MAX, pmesh.GetComm());
+      REQUIRE(norm_g == MFEM_Approx(0.0));
+      MPI_Barrier(MPI_COMM_WORLD);
+   }
```
Comment (helloworld922) [inline tests/unit/dfem/test_mass.cpp:158]: Add a test for `GetDerivative` on the Mass operator applied to the boundary

Comment (jandrej) [reply tests/unit/dfem/test_mass.cpp:158]: added

File: fem/dfem/doperator.hpp (hunk 1 above)
Comment (jandrej) [reply fem/dfem/doperator.hpp]: Are you insisting on that change or would you be OK with the current logic?

Comment (helloworld922) [reply fem/dfem/doperator.hpp]: It doesn't necessarily have to be as part of this PR, but requiring a CPU O(N) operation and extra storage on all elements per integrator seems wasteful when it can be cached inside the mesh and re-used between integrators. It looks like actions/derivative actions also capture this by value so it's duplicated even more times within a single integrator?

Comment (jandrej) [reply fem/dfem/doperator.hpp]: Yea it's captured as a copy once per integrator. Let me try something that mostly resembles your suggestion tomorrow. The number of operations is irrelevant in the grand scheme of things. I agree on the storage part though.

File: fem/dfem/doperator.hpp [hunk 3]
```diff
@@ -477,32 +563,44 @@ void DifferentiableOperator::AddDomainIntegrator(
...
-
-   Array<int> elem_attributes;
-   elem_attributes.SetSize(mesh.GetNE());
-   for (int i = 0; i < mesh.GetNE(); ++i)
+      elem_attributes = &mesh.GetElementAttributes();
+   }
+   else if constexpr (std::is_same_v<entity_t, Entity::BoundaryElement>)
    {
-      elem_attributes[i] = mesh.GetAttribute(i);
+      elem_attributes = &mesh.GetBdrFaceAttributes();
```
Comment (helloworld922) [inline fem/dfem/doperator.hpp:573]: `mesh.GetBdrFaceAttributes` has extra logic beyond a simple loop over `mesh.GetNBE` gathering `mesh.GetBdrAttribute` to get "boundary faces". Do you know if this is ok or not? It's not entirely clear to me what the distinction between mesh boundary elements vs. boundary faces is, but this is the logic used by bilinear/linear forms.

File: fem/dfem/doperator.hpp [hunk 4]
```diff
@@ -630,7 +743,7 @@ void DifferentiableOperator::AddDomainIntegrator(
          test_vdim,             // int (= output_fop.vdim)
          test_op_dim,           // int (derived from output_fop)
          inputs,                // mfem::future::tuple
-         domain_attributes,     // Array<int>
+         attributes,     // Array<int>
```
Comment (sohailreddy) [inline fem/dfem/doperator.hpp]: Align comment indentation

File: fem/dfem/doperator.hpp [hunk 5]
```diff
@@ -739,7 +852,7 @@ void DifferentiableOperator::AddDomainIntegrator(
                test_vdim,             // int (= output_fop.vdim)
                test_op_dim,           // int (derived from output_fop)
                inputs,                // mfem::future::tuple
-               domain_attributes,     // Array<int>
+               attributes,     // Array<int>
```
Comment (sohailreddy) [inline fem/dfem/doperator.hpp]: Align comment indentation

File: fem/dfem/qfunction_apply.hpp [hunk 6]
```diff
@@ -164,6 +173,21 @@ void call_qfunction_derivative_action(
             }
          }
       }
+      else
```
Comment (sohailreddy) [inline fem/dfem/qfunction_apply.hpp:191]: It's better to check "else if (dimension == 1)", and then "else" to abort/do-nothing ?

File: fem/dfem/interpolate.hpp [hunk 7]
```diff
@@ -547,7 +645,14 @@ void map_direction_to_quadrature_data_conditional(
       {
          if (use_sum_factorization)
          {
-            if (dimension == 2)
+            if (dimension == 1)
+            {
+               map_field_to_quadrature_data_tensor_product_1d(
+                  directions_qp[i], dtqmaps[i], direction_e, get<i>(fops),
+                  integration_weights, scratch_mem);
+
```
Comment (sohailreddy) [inline fem/dfem/interpolate.hpp]: delete empty line at 653

File: linalg/tensor.hpp [hunk 8]
```diff
@@ -1243,6 +1243,27 @@ T norm(const tensor<T, n...>& A)
...
+   {
+      return norm(A);
+   }
+   else if constexpr ((n == 3) && (m == 2))
+   {
+      T E = A[0][0] * A[0][0] + A[1][0] * A[1][0] + A[2][0] * A[2][0];
+      T G = A[0][1] * A[0][1] + A[1][1] * A[1][1] + A[2][1] * A[2][1];
+      T F = A[0][0] * A[0][1] + A[1][0] * A[1][1] + A[2][0] * A[2][1];
+      return std::sqrt(E * G - F * F);
+   }
```
Comment (sohailreddy) [inline linalg/tensor.hpp:1266]: safer to either abort after last "else if" or return (dummy?) value.

**Review by sohailreddy — CHANGES_REQUESTED** [review]
minor styling changes and some safer conditional checks.

File: linalg/tensor.hpp (hunk 8 above)
Comment (jandrej) [reply linalg/tensor.hpp:1266]: I added a static assert for supported cases

**Review by sohailreddy — APPROVED** [review]

**Review by helloworld922 — APPROVED** [review]

File: tests/unit/dfem/test_mass.cpp [hunk 9]
```diff
@@ -51,37 +51,114 @@ void DFemMass(const char *filename, int p, const int r)
...
+         norm_l = Y.Normlinf();
+         MPI_Allreduce(&norm_l, &norm_g, 1, MPI_DOUBLE, MPI_MAX, pmesh.GetComm());
+         REQUIRE(norm_g == MFEM_Approx(0.0));
+         MPI_Barrier(MPI_COMM_WORLD);
+      }
+   }
 }
 
-TEST_CASE("DFEM Mass", "[Parallel][DFEM]")
+TEST_CASE("DFEM Mass", "[Parallel][dFEM Mass]")
```
Comment (camierjs) [inline tests/unit/dfem/test_mass.cpp]:
```suggestion
TEST_CASE("dFEM Mass", "[Parallel][dFEM]")
```
We should synchronize with the updated tags from #5022

**Comment by tzanio** [comment]
Merged in `next` for testing...

**Comment by tzanio** [comment]
Re-merged in `next` for testing...

File: .clang-format
Comment (pazner) [inline .clang-format:1]: @jandrej @tzanio, looks like this file was committed by mistake

Comment (tzanio) [reply .clang-format:1]: @jandrej, can you please remove it, if that is the case

Comment (jandrej) [reply .clang-format:1]: removed, PR can be squash merged

**Comment by tzanio** [comment]
Re-merged in `next` for testing...

**Comment by v-dobrev** [comment]
Re-merged in `next` for testing...
